### ISSUE TITLE
Objects Pod: exception when watch gets interrupted, leaves defunct processes behind

### ISSUE THREAD

**Issue by gnarc** [splunk/splunk-connect-for-kubernetes#628]
**What happened**:
out of the box fluentd in objects pod is configured to open a "watch" to the K8s API (https://kubernetes.io/docs/reference/using-api/api-concepts/)
```
# oc get cm splunk-connect-splunk-kubernetes-objects -o yaml | less
[...]
      <watch>
        resource_name events
      </watch>
[...] 
```
When this watch gets interrupted by Loadbalancer timeouts, the ruby process will throw an exception
```
2021-06-25 13:32:46 +0000 [warn]: #0 thread exited by unexpected error plugin=Fluent::Plugin::KubernetesObjectsInput title=:watch_events error_class=HTTP::ConnectionError error="error reading from socket: Could not parse data"
#<Thread:0x000055f1dc976980@watch_events /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/plugin_helper/thread.rb:70 run> terminated with exception (report_on_exception is true):
/usr/share/gems/gems/http-4.4.1/lib/http/response/parser.rb:31:in `add': error reading from socket: Could not parse data (HTTP::ConnectionError)
        from /usr/share/gems/gems/http-4.4.1/lib/http/connection.rb:214:in `read_more'
        from /usr/share/gems/gems/http-4.4.1/lib/http/connection.rb:92:in `readpartial'
        from /usr/share/gems/gems/http-4.4.1/lib/http/response/body.rb:30:in `readpartial'
        from /usr/share/gems/gems/http-4.4.1/lib/http/response/body.rb:36:in `each'
        from /usr/share/gems/gems/kubeclient-4.6.0/lib/kubeclient/watch_stream.rb:25:in `each'
        from /opt/app-root/src/gem/fluent-plugin-kubernetes-objects-1.1.5/lib/fluent/plugin/in_kubernetes_objects.rb:248:in `block (2 levels) in create_watcher_thread'
        from /opt/app-root/src/gem/fluent-plugin-kubernetes-objects-1.1.5/lib/fluent/plugin/in_kubernetes_objects.rb:246:in `tap'
        from /opt/app-root/src/gem/fluent-plugin-kubernetes-objects-1.1.5/lib/fluent/plugin/in_kubernetes_objects.rb:246:in `block in create_watcher_thread'
        from /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
/usr/share/gems/gems/http-4.4.1/lib/http/response/parser.rb:31:in `add': Could not parse data (IOError)
        from /usr/share/gems/gems/http-4.4.1/lib/http/connection.rb:214:in `read_more'
        from /usr/share/gems/gems/http-4.4.1/lib/http/connection.rb:92:in `readpartial'
        from /usr/share/gems/gems/http-4.4.1/lib/http/response/body.rb:30:in `readpartial'
        from /usr/share/gems/gems/http-4.4.1/lib/http/response/body.rb:36:in `each'
        from /usr/share/gems/gems/kubeclient-4.6.0/lib/kubeclient/watch_stream.rb:25:in `each'
        from /opt/app-root/src/gem/fluent-plugin-kubernetes-objects-1.1.5/lib/fluent/plugin/in_kubernetes_objects.rb:248:in `block (2 levels) in create_watcher_thread'
        from /opt/app-root/src/gem/fluent-plugin-kubernetes-objects-1.1.5/lib/fluent/plugin/in_kubernetes_objects.rb:246:in `tap'
        from /opt/app-root/src/gem/fluent-plugin-kubernetes-objects-1.1.5/lib/fluent/plugin/in_kubernetes_objects.rb:246:in `block in create_watcher_thread'
        from /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
2021-06-25 13:32:46 +0000 [error]: #0 unexpected error error_class=HTTP::ConnectionError error="error reading from socket: Could not parse data"
  2021-06-25 13:32:46 +0000 [error]: #0 /usr/share/gems/gems/http-4.4.1/lib/http/response/parser.rb:31:in `add'
  2021-06-25 13:32:46 +0000 [error]: #0 /usr/share/gems/gems/http-4.4.1/lib/http/connection.rb:214:in `read_more'
  2021-06-25 13:32:46 +0000 [error]: #0 /usr/share/gems/gems/http-4.4.1/lib/http/connection.rb:92:in `readpartial'
  2021-06-25 13:32:46 +0000 [error]: #0 /usr/share/gems/gems/http-4.4.1/lib/http/response/body.rb:30:in `readpartial'
  2021-06-25 13:32:46 +0000 [error]: #0 /usr/share/gems/gems/http-4.4.1/lib/http/response/body.rb:36:in `each'
  2021-06-25 13:32:46 +0000 [error]: #0 /usr/share/gems/gems/kubeclient-4.6.0/lib/kubeclient/watch_stream.rb:25:in `each'
  2021-06-25 13:32:46 +0000 [error]: #0 /opt/app-root/src/gem/fluent-plugin-kubernetes-objects-1.1.5/lib/fluent/plugin/in_kubernetes_objects.rb:248:in `block (2 levels) in create_watcher_thread'
  2021-06-25 13:32:46 +0000 [error]: #0 /opt/app-root/src/gem/fluent-plugin-kubernetes-objects-1.1.5/lib/fluent/plugin/in_kubernetes_objects.rb:246:in `tap'
  2021-06-25 13:32:46 +0000 [error]: #0 /opt/app-root/src/gem/fluent-plugin-kubernetes-objects-1.1.5/lib/fluent/plugin/in_kubernetes_objects.rb:246:in `block in create_watcher_thread'
  2021-06-25 13:32:46 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/plugin_helper/thread.rb:78:in `block in thread_create'
2021-06-25 13:32:46 +0000 [error]: #0 unexpected error error_class=HTTP::ConnectionError error="error reading from socket: Could not parse data"
  2021-06-25 13:32:46 +0000 [error]: #0 suppressed same stacktrace
2021-06-25 13:32:46 +0000 [info]: Worker 0 finished unexpectedly with status 1
2021-06-25 13:32:47 +0000 [info]: adding filter pattern="kube.**" type="jq_transformer"
2021-06-25 13:32:47 +0000 [info]: adding filter pattern="kube.**" type="jq_transformer"
2021-06-25 13:32:47 +0000 [info]: adding match pattern="kube.**" type="splunk_hec"
2021-06-25 13:32:47 +0000 [info]: adding source type="monitor_agent"
2021-06-25 13:32:47 +0000 [info]: adding source type="kubernetes_objects"
2021-06-25 13:32:48 +0000 [warn]: #0 both of Plugin @id and path for <storage> are not specified. Using on-memory store.
/usr/share/gems/gems/kubeclient-4.6.0/lib/kubeclient.rb:27: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/usr/share/gems/gems/kubeclient-4.6.0/lib/kubeclient/common.rb:61: warning: The called method `initialize_client' is defined here
2021-06-25 13:32:48 +0000 [info]: adding source type="kubernetes_objects"
2021-06-25 13:32:48 +0000 [warn]: #0 both of Plugin @id and path for <storage> are not specified. Using on-memory store.
2021-06-25 13:32:48 +0000 [warn]: parameter 'cluster_name' in <fields>
  cluster_name
</fields> is not used.
2021-06-25 13:32:48 +0000 [info]: #0 starting fluentd worker pid=784 ppid=1 worker=0
/usr/share/gems/gems/fluentd-1.11.5/lib/fluent/plugin_helper/http_server/compat/server.rb:84: warning: Using the last argument as keyword parameters is deprecated; maybe ** should be added to the call
/usr/share/gems/gems/fluentd-1.11.5/lib/fluent/plugin_helper/http_server/compat/webrick_handler.rb:26: warning: The called method `build' is defined here
2021-06-25 13:32:48 +0000 [info]: #0 fluentd worker is now running worker=0
```
The process restarts and is working again, but will leave behind a defunct process
When this has happend often enough the container cannot fork any new processes and will fail to work

```
`2021-06-25 10:19:17 +0000 [info]: #0 starting fluentd worker pid=48465 ppid=1 worker=0
sh: fork: retry: Resource temporarily unavailable
sh: fork: retry: Resource temporarily unavailable
sh: fork: retry: Resource temporarily unavailable
2021-06-25 10:19:20 +0000 [error]: #0 unexpected error error_class=ThreadError error="can't create Thread: Resource temporarily unavailable"
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/plugin_helper/thread.rb:70:in `initialize'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/plugin_helper/thread.rb:70:in `new'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/plugin_helper/thread.rb:70:in `thread_create'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/plugin/output.rb:461:in `block in start'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/plugin/output.rb:458:in `times'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/plugin/output.rb:458:in `start'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/compat/output.rb:392:in `start'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/plugin_helper/formatter.rb:96:in `start'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluent-plugin-splunk-hec-1.2.5/lib/fluent/plugin/out_splunk_hec.rb:141:in `start'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/compat/call_super_mixin.rb:42:in `start'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/root_agent.rb:200:in `block in start'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/root_agent.rb:189:in `block (2 levels) in lifecycle'
2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/root_agent.rb:188:in `each'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/root_agent.rb:188:in `block in lifecycle'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/root_agent.rb:175:in `each'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/root_agent.rb:175:in `lifecycle'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/root_agent.rb:199:in `start'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/engine.rb:248:in `start'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/engine.rb:147:in `run'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/supervisor.rb:607:in `block in run_worker'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/supervisor.rb:845:in `main_process'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/supervisor.rb:598:in `run_worker'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/command/fluentd.rb:361:in `<top (required)>'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/bin/fluentd:8:in `require'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/share/gems/gems/fluentd-1.11.5/bin/fluentd:8:in `<top (required)>'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/bin/fluentd:23:in `load'
  2021-06-25 10:19:20 +0000 [error]: #0 /usr/bin/fluentd:23:in `<main>'
2021-06-25 10:19:20 +0000 [error]: #0 unexpected error error_class=ThreadError error="can't create Thread: Resource temporarily unavailable"
  2021-06-25 10:19:20 +0000 [error]: #0 suppressed same stacktrace
sh: fork: retry: Resource temporarily unavailable
2021-06-25 10:19:20 +0000 [info]: Worker 0 finished unexpectedly with status 1
2021-06-25 10:19:22 +0000 [info]: adding filter pattern="kube.**" type="jq_transformer"
sh: fork: retry: Resource temporarily unavailable
sh: fork: Resource temporarily unavailable
Unexpected error can't create Thread: Resource temporarily unavailable
  /usr/share/gems/gems/serverengine-2.2.3/lib/serverengine/signal_thread.rb:71:in `initialize'
  /usr/share/gems/gems/serverengine-2.2.3/lib/serverengine/signal_thread.rb:71:in `new'
  /usr/share/gems/gems/serverengine-2.2.3/lib/serverengine/signal_thread.rb:71:in `signal_handler_main'
  /usr/share/gems/gems/serverengine-2.2.3/lib/serverengine/signal_thread.rb:44:in `block in trap'
  /usr/share/gems/gems/serverengine-2.2.3/lib/serverengine/process_manager.rb:255:in `select'
  /usr/share/gems/gems/serverengine-2.2.3/lib/serverengine/process_manager.rb:255:in `tick'
  /usr/share/gems/gems/serverengine-2.2.3/lib/serverengine/multi_spawn_server.rb:84:in `wait_tick'
  /usr/share/gems/gems/serverengine-2.2.3/lib/serverengine/multi_worker_server.rb:60:in `run'
  /usr/share/gems/gems/serverengine-2.2.3/lib/serverengine/multi_spawn_server.rb:50:in `run'
  /usr/share/gems/gems/serverengine-2.2.3/lib/serverengine/server.rb:128:in `main'
  /usr/share/gems/gems/serverengine-2.2.3/lib/serverengine/daemon.rb:119:in `main'
  /usr/share/gems/gems/serverengine-2.2.3/lib/serverengine/daemon.rb:68:in `run'
  /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/supervisor.rb:717:in `supervise'
  /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/supervisor.rb:566:in `run_supervisor'
  /usr/share/gems/gems/fluentd-1.11.5/lib/fluent/command/fluentd.rb:341:in `<top (required)>'
  /usr/share/gems/gems/fluentd-1.11.5/bin/fluentd:8:in `require'
  /usr/share/gems/gems/fluentd-1.11.5/bin/fluentd:8:in `<top (required)>'
  /usr/bin/fluentd:23:in `load'
  /usr/bin/fluentd:23:in `<top (required)>'
  /usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:63:in `load'
  /usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:63:in `kernel_load'
  /usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli/exec.rb:28:in `run'
  /usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli.rb:476:in `exec'
  /usr/share/gems/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
  /usr/share/gems/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
  /usr/share/gems/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor.rb:399:in `dispatch'
  /usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli.rb:30:in `dispatch'
  /usr/share/gems/gems/bundler-2.1.4/lib/bundler/vendor/thor/lib/thor/base.rb:476:in `start'
  /usr/share/gems/gems/bundler-2.1.4/lib/bundler/cli.rb:24:in `start'
  /usr/share/gems/gems/bundler-2.1.4/libexec/bundle:46:in `block in <top (required)>'
  /usr/share/gems/gems/bundler-2.1.4/lib/bundler/friendly_errors.rb:123:in `with_friendly_errors'
  /usr/share/gems/gems/bundler-2.1.4/libexec/bundle:34:in `<top (required)>'
  /usr/local/bin/bundle:23:in `load'
  /usr/local/bin/bundle:23:in `<main>'

# oc rsh $(oc get pod -o name -l app=splunk-kubernetes-objects) bash -c " ps wwwaux|grep defunct|grep -v grep |wc -l"
696

#oc rsh ...
USER         PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
1000580+       1  0.0  0.4 420092 65708 ?        Ssl  Jun25   2:13 /usr/bin/fluentd -c /fluentd/etc/fluent.conf
1000580+      21  0.0  0.0      0     0 ?        Z    Jun25   0:00 [sh] <defunct>
1000580+      22  0.0  0.0      0     0 ?        Z    Jun25   0:00 [sh] <defunct>
[...]
1000580+   18193  0.0  0.0      0     0 ?        Z    10:41   0:00 [sh] <defunct>
1000580+   18235  3.5  1.4 2028952 233540 ?      Sl   10:51   0:04 /usr/bin/ruby -r/usr/share/gems/gems/bundler-2.1.4/lib/bundler/setup -Eascii-8bit:ascii-8bit /usr/bin/fluentd -c /fluentd/etc/fluent.conf --under-supervisor
1000580+   18241  0.0  0.0  14788  3500 ?        S    10:51   0:00 sh -c jq  --unbuffered -c '.record.cluster_name = "ocgiax03" | .record' 2>&1
1000580+   18242  0.0  0.0  14788  3516 ?        S    10:51   0:00 sh -c jq  --unbuffered -c '.record.source = "namespace:\(env.MY_NAMESPACE)/pod:\(env.MY_POD_NAME)" | .record.sourcetype = (.tag | gsub("\\."; ":")) | .record' 2>&1
1000580+   18248  0.7  0.0  16692  3756 ?        S    10:51   0:00 jq --unbuffered -c .record.cluster_name = "ocgiax03" | .record
1000580+   18249  0.8  0.0  16592  3928 ?        S    10:51   0:01 jq --unbuffered -c .record.source = "namespace:\(env.MY_NAMESPACE)/pod:\(env.MY_POD_NAME)" | .record.sourcetype = (.tag | gsub("\\."; ":")) | .record
1000580+   18284  0.0  0.0  14920  4028 pts/0    Ss   10:52   0:00 /bin/sh
1000580+   18291  0.0  0.0  47536  4164 pts/0    R+   10:53   0:00 ps wwwaux
1000580+   18292  0.0  0.0   9504  2164 pts/0    S+   10:53   0:00 more

```
The reason this happens in Shiftstack (OpenShift on Openstack) with kuryr CNI is because the service the objects pod process is connecting to is under the hood an Amphora LB (Haproxy) wich comes with timeout values of 10min for frontend and backend connections. This configuration will cut any connection after 10 min if nothing happend until then (events, deployments, ...whatever you decided to watch) and the process will execpt, restart and leave the defunct process behind. If we raise the timeouts to say one day, the problem almost disappears

**What you expected to happen**:
I would expect the process to better act on interrupted connections w/o leaving defunct processes behind.

**How to reproduce it (as minimally and precisely as possible)**:
Unless you find a way to set timeouts for the internal API service in the OpenShift default project you need Openstack with kuryr on a cluster where not much is happening. Or you could force it with values <10min and 1d to see that the problem goes away.
```   
    oc get all -n default
    openstack loadbalancer list  → look for amphora lb with corresponding ClusterIP
    openstack loadbalancer listener list → finder listener to that LB
    openstack loadbalancer listener show mylittle-kuryr-api-loadbalancer-listener
compare values and set e.g. to one day:
     timeout_client_data |86400000
     timeout_member_data |86400000
This is disruptive and may create some noise in OCP

Check the logs after some while and the process list within the SCK-objects container.
```

**Anything else we need to know?**:
WORKAROUNDS
1. One workaround is described above: change the timeout values; we refrained from this because of administrative overhead, possible sideeffects and because we believe this should be handled in the Objects-Pod-Code.
2. Do nothing. Let the process die and a liveness probe (e.g. https://github.com/splunk/splunk-connect-for-kubernetes/issues/626) handle the hanging pod once it runs into fork issues.

**Environment**:
Openshift running on Openstack with kuryr CNI
- Kubernetes version (use `kubectl version`):
```
# oc version
Client Version: 4.6.32
Server Version: 4.6.32
Kubernetes Version: v1.19.0+d670f74
```
- Splunk Connect for Kubernetes helm chart version: 1.4.3, 1.4.7

**Comment by rockb1017** [comment]
I have added error handling so that it would just recreate watch connection. Could you try with this image and see how it is? `rock1017/kube-objects:1.1.7rc`

**Comment by gnarc** [comment]
Hi Rock, 

thanks for that.
Rolled it out on  two OCP clusters and it is looking good:
`
Cluster 1:
2021-07-20 08:03:09 +0000 [warn]: #0 both of Plugin @id and path for <storage> are not specified. Using on-memory store.
2021-07-20 08:03:09 +0000 [info]: #0 starting fluentd worker pid=15 ppid=1 worker=0
2021-07-20 08:03:09 +0000 [info]: #0 fluentd worker is now running worker=0
2021-07-20 08:13:09 +0000 [info]: #0 Got exception error reading from socket: Could not parse data. Resetting eatcher.
2021-07-20 08:23:09 +0000 [info]: #0 Got exception error reading from socket: Could not parse data. Resetting eatcher.
2021-07-20 08:33:09 +0000 [info]: #0 Got exception error reading from socket: Could not parse data. Resetting eatcher.

Cluster 2:
2021-07-20 08:11:23 +0000 [info]: #0 starting fluentd worker pid=15 ppid=1 worker=0
2021-07-20 08:11:23 +0000 [info]: #0 fluentd worker is now running worker=0
2021-07-20 08:21:24 +0000 [info]: #0 Got exception error reading from socket: Could not parse data. Resetting eatcher.
2021-07-20 08:31:24 +0000 [info]: #0 Got exception error reading from socket: Could not parse data. Resetting eatcher.
2021-07-20 08:31:24 +0000 [info]: #0 Got exception error reading from socket: Could not parse data. Resetting eatcher.
2021-07-20 08:41:24 +0000 [info]: #0 Got exception error reading from socket: Could not parse data. Resetting eatcher.

`
The defunct processes seem to be history too.

You might want to change "eatcher" to "watcher" unless there is a pun intended.

If I observe anything strange in the next 1-2 weeks I will drop you a line


cheers

**Comment by gnarc** [comment]
Testing was a bit chaotic with different versions from 1.4.3 to 1.4.8.
In general it seems to work. I had once an issue when the process no longer worked after a couple of days, i.e. no events were sent to the backend w/o anything obvious.
The pod/container was fine because it logged the kill signal from `oc delete pod ...`. 

Today I deployed it on a couple of machines with SCK 1.4.9 and a custom liveness probe. The one provided would not cover an issue like this and since the container does no longer "repair" itself by dying I took the one from https://github.com/splunk/splunk-connect-for-kubernetes/issues/493 and extended the periodSeconds: from 60s to 300s which should make it useable in our environment.
